File: scripts/backfill_embeddings.py
#!/usr/bin/env python3
"""
backfill_embeddings.py — Backfill embeddings for existing helm_beliefs and helm_entities rows.

Queries all rows with embedding IS NULL, generates text-embedding-3-small vectors
via the OpenAI API, and PATCHes each row. Idempotent — safe to re-run.

Usage:
    python3 scripts/backfill_embeddings.py [--dry-run]

Environment variables required:
    OPENAI_API_KEY           — OpenAI API key for embedding generation
    SUPABASE_BRAIN_SERVICE_KEY — Supabase service role key

Brain URL and service key env var name are read from hammerfall-config.md.
"""

import argparse
import os
import re
import sys
import time
import urllib.error
import urllib.request
import json
from pathlib import Path

# ---------------------------------------------------------------------------
# Config
# ---------------------------------------------------------------------------

EMBEDDING_MODEL = "text-embedding-3-small"
BATCH_SIZE = 20          # rows per batch before sleeping
BATCH_SLEEP = 1.0        # seconds between batches (rate limit headroom)
REQUEST_TIMEOUT = 20     # seconds per OpenAI API call

SCRIPT_DIR = Path(__file__).parent
CONFIG_FILE = SCRIPT_DIR.parent / "hammerfall-config.md"


def read_config() -> tuple[str, str]:
    """Read BRAIN_URL and SERVICE_KEY from hammerfall-config.md + env."""
    text = CONFIG_FILE.read_text()

    url_match = re.search(r"supabase_brain_url:\s*(\S+)", text)
    key_env_match = re.search(r"supabase_brain_service_key_env:\s*(\S+)", text)

    if not url_match:
        sys.exit("ERROR: supabase_brain_url not found in hammerfall-config.md")
    if not key_env_match:
        sys.exit("ERROR: supabase_brain_service_key_env not found in hammerfall-config.md")

    brain_url = url_match.group(1).rstrip("/")
    key_env = key_env_match.group(1)
    service_key = os.environ.get(key_env)

    if not service_key:
        sys.exit(f"ERROR: env var '{key_env}' is not set.")

    return brain_url, service_key


def get_openai_key() -> str:
    key = os.environ.get("OPENAI_API_KEY")
    if not key:
        sys.exit("ERROR: OPENAI_API_KEY is not set.")
    return key


# ---------------------------------------------------------------------------
# API helpers
# ---------------------------------------------------------------------------

def supabase_get(brain_url: str, service_key: str, path: str) -> list:
    """GET from Supabase REST API."""
    req = urllib.request.Request(
        f"{brain_url}/rest/v1/{path}",
        headers={
            "apikey": service_key,
            "Authorization": f"Bearer {service_key}",
        },
    )
    with urllib.request.urlopen(req, timeout=15) as resp:
        return json.load(resp)


def supabase_patch(brain_url: str, service_key: str, table: str, row_id: str, payload: dict) -> None:
    """PATCH a single row by id."""
    data = json.dumps(payload).encode()
    req = urllib.request.Request(
        f"{brain_url}/rest/v1/{table}?id=eq.{row_id}",
        data=data,
        headers={
            "apikey": service_key,
            "Authorization": f"Bearer {service_key}",
            "Content-Type": "application/json",
            "Prefer": "return=minimal",
        },
        method="PATCH",
    )
    with urllib.request.urlopen(req, timeout=15) as resp:
        resp.read()


def generate_embedding(text: str, openai_key: str) -> list[float] | None:
    """Call text-embedding-3-small. Returns vector or None on failure."""
    data = json.dumps({"model": EMBEDDING_MODEL, "input": text}).encode()
    req = urllib.request.Request(
        "https://api.openai.com/v1/embeddings",
        data=data,
        headers={
            "Authorization": f"Bearer {openai_key}",
            "Content-Type": "application/json",
        },
        method="POST",
    )
    try:
        with urllib.request.urlopen(req, timeout=REQUEST_TIMEOUT) as resp:
            result = json.load(resp)
            return result["data"][0]["embedding"]
    except Exception as e:
        print(f"    WARNING: embedding generation failed — {e}")
        return None


# ---------------------------------------------------------------------------
# Table backfills
# ---------------------------------------------------------------------------

def backfill_beliefs(brain_url: str, service_key: str, openai_key: str, dry_run: bool) -> tuple[int, int]:
    """
    Backfill helm_beliefs rows with embedding IS NULL.
    Embed the belief text (the main semantic content).
    Returns (success_count, failed_count).
    """
    print("\n--- helm_beliefs ---")
    rows = supabase_get(
        brain_url, service_key,
        "helm_beliefs?select=id,belief&embedding=is.null&active=eq.true",
    )
    print(f"  {len(rows)} rows with null embedding")

    if not rows:
        return 0, 0

    success = 0
    failed = 0

    for i, row in enumerate(rows):
        row_id = row["id"]
        belief_text = row.get("belief", "").strip()

        if not belief_text:
            print(f"  SKIP  [{i+1}/{len(rows)}] id={row_id} — empty belief text")
            failed += 1
            continue

        print(f"  [{i+1}/{len(rows)}] id={row_id} — {belief_text[:60]}...")

        if dry_run:
            print("    DRY RUN — skipping embed + patch")
            success += 1
        else:
            embedding = generate_embedding(belief_text, openai_key)
            if embedding is None:
                failed += 1
            else:
                try:
                    supabase_patch(brain_url, service_key, "helm_beliefs", row_id, {"embedding": embedding})
                    success += 1
                except Exception as e:
                    print(f"    ERROR: patch failed — {e}")
                    failed += 1

        if (i + 1) % BATCH_SIZE == 0 and not dry_run:
            print(f"  Batch complete — sleeping {BATCH_SLEEP}s")
            time.sleep(BATCH_SLEEP)

    return success, failed


def backfill_entities(brain_url: str, service_key: str, openai_key: str, dry_run: bool) -> tuple[int, int]:
    """
    Backfill helm_entities rows with embedding IS NULL.
    Embed "name — summary" when summary is present, else just name.
    Returns (success_count, failed_count).
    """
    print("\n--- helm_entities ---")
    rows = supabase_get(
        brain_url, service_key,
        "helm_entities?select=id,name,summary&embedding=is.null&active=eq.true",
    )
    print(f"  {len(rows)} rows with null embedding")

    if not rows:
        return 0, 0

    success = 0
    failed = 0

    for i, row in enumerate(rows):
        row_id = row["id"]
        name = (row.get("name") or "").strip()
        summary = (row.get("summary") or "").strip()

        if not name:
            print(f"  SKIP  [{i+1}/{len(rows)}] id={row_id} — empty name")
            failed += 1
            continue

        embed_text = f"{name} — {summary}" if summary else name
        print(f"  [{i+1}/{len(rows)}] id={row_id} — {embed_text[:60]}...")

        if dry_run:
            print("    DRY RUN — skipping embed + patch")
            success += 1
        else:
            embedding = generate_embedding(embed_text, openai_key)
            if embedding is None:
                failed += 1
            else:
                try:
                    supabase_patch(brain_url, service_key, "helm_entities", row_id, {"embedding": embedding})
                    success += 1
                except Exception as e:
                    print(f"    ERROR: patch failed — {e}")
                    failed += 1

        if (i + 1) % BATCH_SIZE == 0 and not dry_run:
            print(f"  Batch complete — sleeping {BATCH_SLEEP}s")
            time.sleep(BATCH_SLEEP)

    return success, failed


# ---------------------------------------------------------------------------
# Entry point
# ---------------------------------------------------------------------------

def main():
    parser = argparse.ArgumentParser(description="Backfill embeddings for helm_beliefs and helm_entities.")
    parser.add_argument("--dry-run", action="store_true", help="Query rows and print plan without writing.")
    args = parser.parse_args()

    if args.dry_run:
        print("=== DRY RUN — no writes will occur ===")

    brain_url, service_key = read_config()
    openai_key = get_openai_key() if not args.dry_run else "dry-run"

    print(f"Brain URL: {brain_url}")
    print(f"Model:     {EMBEDDING_MODEL}")

    b_success, b_failed = backfill_beliefs(brain_url, service_key, openai_key, args.dry_run)
    e_success, e_failed = backfill_entities(brain_url, service_key, openai_key, args.dry_run)

    total_success = b_success + e_success
    total_failed = b_failed + e_failed

    print(f"\n=== Backfill complete ===")
    print(f"  helm_beliefs:  {b_success} embedded, {b_failed} failed/skipped")
    print(f"  helm_entities: {e_success} embedded, {e_failed} failed/skipped")
    print(f"  Total:         {total_success} embedded, {total_failed} failed/skipped")

    if total_failed > 0:
        print(f"\n  {total_failed} rows still have null embedding — re-run to retry.")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: scripts/backfill_embeddings.js
#!/usr/bin/env node
/**
 * backfill_embeddings.js — Backfill embeddings for existing helm_beliefs and helm_entities rows.
 *
 * Queries all active rows with embedding IS NULL, generates text-embedding-3-small
 * vectors via the OpenAI API, and PATCHes each row. Idempotent — safe to re-run.
 *
 * Usage:
 *   node scripts/backfill_embeddings.js [--dry-run]
 *
 * Environment variables required:
 *   OPENAI_API_KEY             — OpenAI API key
 *   SUPABASE_BRAIN_SERVICE_KEY — Supabase service role key
 *
 * Brain URL is read from hammerfall-config.md.
 */

import { readFileSync } from "fs";
import { join, dirname } from "path";
import { fileURLToPath } from "url";

const __dirname = dirname(fileURLToPath(import.meta.url));

const EMBEDDING_MODEL = "text-embedding-3-small";
const BATCH_SIZE = 20;
const BATCH_SLEEP_MS = 1000;
const DRY_RUN = process.argv.includes("--dry-run");

// ---------------------------------------------------------------------------
// Config
// ---------------------------------------------------------------------------

function readConfig() {
  const configPath = join(__dirname, "..", "hammerfall-config.md");
  const text = readFileSync(configPath, "utf8");

  const urlMatch = text.match(/supabase_brain_url:\s*(\S+)/);
  const keyEnvMatch = text.match(/supabase_brain_service_key_env:\s*(\S+)/);

  if (!urlMatch) throw new Error("supabase_brain_url not found in hammerfall-config.md");
  if (!keyEnvMatch) throw new Error("supabase_brain_service_key_env not found in hammerfall-config.md");

  const brainUrl = urlMatch[1].replace(/\/$/, "");
  const keyEnv = keyEnvMatch[1];
  const serviceKey = process.env[keyEnv];

  if (!serviceKey) throw new Error(`Env var '${keyEnv}' is not set.`);

  return { brainUrl, serviceKey };
}

function getOpenAIKey() {
  const key = process.env.OPENAI_API_KEY;
  if (!key) throw new Error("OPENAI_API_KEY is not set.");
  return key;
}

// ---------------------------------------------------------------------------
// API helpers
// ---------------------------------------------------------------------------

async function supabaseGet(brainUrl, serviceKey, path) {
  const res = await fetch(`${brainUrl}/rest/v1/${path}`, {
    headers: {
      apikey: serviceKey,
      Authorization: `Bearer ${serviceKey}`,
    },
  });
  if (!res.ok) throw new Error(`Supabase GET failed: ${res.status} ${await res.text()}`);
  return res.json();
}

async function supabasePatch(brainUrl, serviceKey, table, rowId, payload) {
  const res = await fetch(`${brainUrl}/rest/v1/${table}?id=eq.${rowId}`, {
    method: "PATCH",
    headers: {
      apikey: serviceKey,
      Authorization: `Bearer ${serviceKey}`,
      "Content-Type": "application/json",
      Prefer: "return=minimal",
    },
    body: JSON.stringify(payload),
  });
  if (!res.ok) throw new Error(`Supabase PATCH failed: ${res.status} ${await res.text()}`);
}

async function generateEmbedding(text, openaiKey) {
  const res = await fetch("https://api.openai.com/v1/embeddings", {
    method: "POST",
    headers: {
      Authorization: `Bearer ${openaiKey}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ model: EMBEDDING_MODEL, input: text }),
  });
  if (!res.ok) throw new Error(`OpenAI embeddings failed: ${res.status} ${await res.text()}`);
  const data = await res.json();
  return data.data[0].embedding;
}

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

// ---------------------------------------------------------------------------
// Table backfills
// ---------------------------------------------------------------------------

async function backfillBeliefs(brainUrl, serviceKey, openaiKey) {
  console.log("\n--- helm_beliefs ---");
  const rows = await supabaseGet(
    brainUrl, serviceKey,
    "helm_beliefs?select=id,belief&embedding=is.null&active=eq.true"
  );
  console.log(`  ${rows.length} rows with null embedding`);
  if (!rows.length) return { success: 0, failed: 0 };

  let success = 0, failed = 0;

  for (let i = 0; i < rows.length; i++) {
    const { id, belief } = rows[i];
    const text = (belief || "").trim();

    if (!text) {
      console.log(`  SKIP  [${i+1}/${rows.length}] id=${id} — empty belief text`);
      failed++;
      continue;
    }

    console.log(`  [${i+1}/${rows.length}] id=${id} — ${text.slice(0, 60)}...`);

    if (DRY_RUN) {
      console.log("    DRY RUN — skipping embed + patch");
      success++;
    } else {
      try {
        const embedding = await generateEmbedding(text, openaiKey);
        await supabasePatch(brainUrl, serviceKey, "helm_beliefs", id, { embedding });
        success++;
      } catch (e) {
        console.log(`    ERROR: ${e.message}`);
        failed++;
      }
    }

    if ((i + 1) % BATCH_SIZE === 0 && !DRY_RUN) {
      console.log(`  Batch complete — sleeping ${BATCH_SLEEP_MS}ms`);
      await sleep(BATCH_SLEEP_MS);
    }
  }

  return { success, failed };
}

async function backfillEntities(brainUrl, serviceKey, openaiKey) {
  console.log("\n--- helm_entities ---");
  const rows = await supabaseGet(
    brainUrl, serviceKey,
    "helm_entities?select=id,name,summary&embedding=is.null&active=eq.true"
  );
  console.log(`  ${rows.length} rows with null embedding`);
  if (!rows.length) return { success: 0, failed: 0 };

  let success = 0, failed = 0;

  for (let i = 0; i < rows.length; i++) {
    const { id, name, summary } = rows[i];
    const nameText = (name || "").trim();
    const summaryText = (summary || "").trim();

    if (!nameText) {
      console.log(`  SKIP  [${i+1}/${rows.length}] id=${id} — empty name`);
      failed++;
      continue;
    }

    const embedText = summaryText ? `${nameText} — ${summaryText}` : nameText;
    console.log(`  [${i+1}/${rows.length}] id=${id} — ${embedText.slice(0, 60)}...`);

    if (DRY_RUN) {
      console.log("    DRY RUN — skipping embed + patch");
      success++;
    } else {
      try {
        const embedding = await generateEmbedding(embedText, openaiKey);
        await supabasePatch(brainUrl, serviceKey, "helm_entities", id, { embedding });
        success++;
      } catch (e) {
        console.log(`    ERROR: ${e.message}`);
        failed++;
      }
    }

    if ((i + 1) % BATCH_SIZE === 0 && !DRY_RUN) {
      console.log(`  Batch complete — sleeping ${BATCH_SLEEP_MS}ms`);
      await sleep(BATCH_SLEEP_MS);
    }
  }

  return { success, failed };
}

async function backfillMemory(brainUrl, serviceKey, openaiKey) {
  console.log("\n--- helm_memory ---");
  const rows = await supabaseGet(
    brainUrl, serviceKey,
    "helm_memory?select=id,content&embedding=is.null"
  );
  console.log(`  ${rows.length} rows with null embedding`);
  if (!rows.length) return { success: 0, failed: 0 };

  let success = 0, failed = 0;

  for (let i = 0; i < rows.length; i++) {
    const { id, content } = rows[i];
    const text = (content || "").trim();

    if (!text) {
      console.log(`  SKIP  [${i+1}/${rows.length}] id=${id} — empty content`);
      failed++;
      continue;
    }

    console.log(`  [${i+1}/${rows.length}] id=${id} — ${text.slice(0, 60)}...`);

    if (DRY_RUN) {
      console.log("    DRY RUN — skipping embed + patch");
      success++;
    } else {
      try {
        const embedding = await generateEmbedding(text, openaiKey);
        await supabasePatch(brainUrl, serviceKey, "helm_memory", id, { embedding });
        success++;
      } catch (e) {
        console.log(`    ERROR: ${e.message}`);
        failed++;
      }
    }

    if ((i + 1) % BATCH_SIZE === 0 && !DRY_RUN) {
      console.log(`  Batch complete — sleeping ${BATCH_SLEEP_MS}ms`);
      await sleep(BATCH_SLEEP_MS);
    }
  }

  return { success, failed };
}

// ---------------------------------------------------------------------------
// Entry point
// ---------------------------------------------------------------------------

async function main() {
  if (DRY_RUN) console.log("=== DRY RUN — no writes will occur ===");

  const { brainUrl, serviceKey } = readConfig();
  const openaiKey = DRY_RUN ? "dry-run" : getOpenAIKey();

  console.log(`Brain URL: ${brainUrl}`);
  console.log(`Model:     ${EMBEDDING_MODEL}`);

  const memory = await backfillMemory(brainUrl, serviceKey, openaiKey);
  const beliefs = await backfillBeliefs(brainUrl, serviceKey, openaiKey);
  const entities = await backfillEntities(brainUrl, serviceKey, openaiKey);

  const totalSuccess = memory.success + beliefs.success + entities.success;
  const totalFailed = memory.failed + beliefs.failed + entities.failed;

  console.log(`\n=== Backfill complete ===`);
  console.log(`  helm_memory:   ${memory.success} embedded, ${memory.failed} failed/skipped`);
  console.log(`  helm_beliefs:  ${beliefs.success} embedded, ${beliefs.failed} failed/skipped`);
  console.log(`  helm_entities: ${entities.success} embedded, ${entities.failed} failed/skipped`);
  console.log(`  Total:         ${totalSuccess} embedded, ${totalFailed} failed/skipped`);

  if (totalFailed > 0) {
    console.log(`\n  ${totalFailed} rows still have null embedding — re-run to retry.`);
    process.exit(1);
  }
}

main().catch((e) => { console.error("FATAL:", e.message); process.exit(1); });
